In [1]:
import pandas as pd

In [2]:
import requests

In [3]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

Using Pandas read html function to extract the relevant data table from above wikipedia page

In [4]:
html = requests.get(url).content
df_list = pd.read_html(html)
df = df_list[0]
print(df)

    Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
282      M8Z         Etobicoke              Mimico NW
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned

[287 rows x 3 columns]


In [5]:
df.shape

(287, 3)

Creating a new dataframe df1 which ignores all the Boroughs marked 'Not assigned'

In [6]:
df1=df[df.Borough!='Not assigned']

In [7]:
df1.shape

(210, 3)

The new dataframe has reduced to 210 rows from 287 rows indicating that 77 rows had borough field Not assigned. This can be checked as follows to count the number of Not assigned Boroughs

In [68]:
NA=df[df['Borough']=='Not assigned']
NA.shape

(77, 3)

In [8]:
df1.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor

Replacing the neigbourhood entries which are 'Not assigned' with name of the corresponding Borough

In [17]:
import numpy as np

In [28]:
df1['Neighbourhood']=np.where(df1['Neighbourhood']=='Not assigned',df1['Borough'],df1['Neighbourhood'])

D:\Users\Sharad\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [64]:
df1.shape

(210, 3)

Creating a new dataframe df2 which has aggregated the all the neigbourhoods with same post codes

In [65]:
df2=df1.groupby(['Postcode','Borough'])['Neighbourhood'].unique().to_frame(name='Neighbourhood').reset_index()

In [66]:
df2.head()

Postcode      Borough                             Neighbourhood
0      M1B  Scarborough                          [Rouge, Malvern]
1      M1C  Scarborough  [Highland Creek, Rouge Hill, Port Union]
2      M1E  Scarborough       [Guildwood, Morningside, West Hill]
3      M1G  Scarborough                                  [Woburn]
4      M1H  Scarborough                               [Cedarbrae]

Checking the shape of dataframe df2

In [67]:
df2.shape

(103, 3)